File: djikstra.cpp
=== Generate Results ===

Here's a C++ implementation of Dijkstra's algorithm with documentation:

```c++
// Graph class representing a graph using adjacency list representation
class Graph {
public:
    // Define a structure for a node in the graph
    struct Node {
        int vertex;   // Vertex number of this node
        int weight;  // Weight of the edge leading to this node from source vertex
        Node* next; // Pointer to the next node in the adjacency list for this vertex
    };

    // Define a structure for the adjacency list representation of the graph
    struct AdjList {
        int vertex;   // Vertex number of this node
        int weight;  // Weight of the edge leading to this node from source vertex
        Node* next; // Pointer to the next node in the adjacency list for this vertex
    };

    // Map representing graph. Key is vertex number, value is AdjList structure
    unordered_map<int, AdjList*> map;
};

// Function to add an edge to the graph
void addEdge(Graph& graph, int src, int dest, int weight) {
    // Create a new node and initialize its properties
    Node* newNode = new Node();
    newNode->vertex = dest;
    newNode->weight = weight;
    
    // If the adjacency list for this vertex doesn't exist, create one and add it to the graph map
    if (graph.map.find(src) == graph.map.end()) {
        AdjList* node = new AdjList();
        node->vertex = src;
        node->next = nullptr;
        graph.map[src] = node;
    }

    // Add the edge to the adjacency list of the source vertex
    Node** head = &(graph.map[src]->next);
    while ((*head) != nullptr && (*head)->vertex < dest) {
        head = &((*head)->next);
    }
    if ((*head) == nullptr || (*head)->vertex > dest) {
        (*head) = newNode;
    } else if ((*head)->weight > weight) { // If the edge already exists and the incoming weight is smaller, update it
        Node* temp = *head;
        *head = newNode;
        newNode->next = temp;
    }
}

// Dijkstra's algorithm implementation for finding shortest path from source vertex to all other vertices in the graph
void dijkstra(Graph& graph, int src) {
    // Initialize distances and predecessors arrays with infinite values and set source distance as 0
    vector<int> distances(graph.map.size(), INT_MAX);
    vector<int> predecessors(graph.map.size(), -1);
    distances[src] = 0;

    // Initialize a priority queue to store the vertices in order of their distance from source vertex
    priority_queue<pair<int, int>, vector<pair<int, int>, greater<pair<int, int>>> > pq;
    for (auto it = graph.map.begin(); it != graph.map.end(); ++it) {
        if (it->first != src) {
            pq.push({ it->second->weight, it->first });
        }
    }

    // Loop until the priority queue is empty
    while (!pq.empty()) {
        int vertex = pq.top().second;
        int distance = pq.top().first;
        pq.pop();

        // Check if we have already processed this vertex or if its distance is greater than the current shortest path
        if (distances[vertex] != INT_MAX && distances[vertex] <= distance) {
            continue;
        }

        // Update the distance of all adjacent vertices and their predecessors
        for (Node* adj = graph.map[vertex]; adj != nullptr; adj = adj->next) {
            int newDistance = distances[vertex] + adj->weight;
            if (distances[adj->vertex] > newDistance) {
                distances[adj->vertex] = newDistance;
                predecessors[adj->vertex] = vertex;

                // Update the priority queue with the new distance of this vertex
                pq.push({ newDistance, adj->vertex });
            }
        }
    }

    // Print the shortest path from source to all other vertices and their distances
    for (int i = 0; i < graph.map.size(); ++i) {
        if (i != src) {
            cout << "Distance from source " << src << " to vertex " << i << ": " << distances[i] << endl;
            int prevVertex = predecessors[i];
            while (prevVertex != -1) {
                cout << "  --> " << prevVertex << endl;
                prevVertex = predecessors[prevVertex];
            }
        }
    }
}
```

File: djikstra.py
# === Generate Results ===

# Here's a C++ implementation of Dijkstra's algorithm with documentation:

# ```python
import heapq
from collections import defaultdict, deque


class Graph:
    def __init__(self):
        self.graph = defaultdict(list)

    def add_edge(self, u, v, w):
        self.graph[u].append((v, w))
        self.graph[v].append((u, w))

    def dijkstra(self, start, end):
        dist = [float("inf")] * len(
            self.graph
        )  # Distance to the start node is initially infinity
        visited = [False] * len(self.graph)  # Initialize all vertices as not visited
        prev = [None] * len(self.graph)
        pq_update = []  # Priority queue for vertices with updated distances
        pq_visited = []  # Priority queue for already visited vertices

        dist[start] = 0  # Set starting node distance to zero
        heapq.heappush(
            pq_update, (0, start)
        )  # Add starting node to update priority queue

        while pq_update or pq_visited:
            # Pop the vertex with the smallest distance from either update or visited priority queues
            if not pq_update:
                dist_v, v = heapq.heappop(pq_visited)
            else:
                dist_v, v = heapq.heappushpop(pq_update, heapq.heappop(pq_update))

            # Mark vertex as visited and add its neighbors to pq_update if shorter path is found
            visited[v] = True
            for neighbor, weight in self.graph[v]:
                dist_neighbor = dist[v] + weight
                if not visited[neighbor] and (
                    not pq_visited or dist_neighbor < min(pq_visited)[0]
                ):
                    heapq.heappush(pq_update, (dist_neighbor, neighbor))
                elif not visited[
                    neighbor
                ]:  # If neighbor is already in pq_visited and has a shorter distance, update it there
                    heapq.heapreplace(pq_visited, (dist_neighbor, neighbor))

            # Move vertices from pq_update to pq_visited when their distance is finalized
            if not visited[v]:
                heapq.heappush(pq_visited, (dist_v, v))

        # Reconstruct shortest path using BFS starting from the end node and return it
        path = deque()
        curr = end
        while prev[curr]:
            path.appendleft(curr)
            curr = prev[curr]
        if start not in path:
            return None, None  # No path found

        path.appendleft(start)
        return list(path), dist[end]
